gitlab/resource_gitlab_project_approval_rule.go

```diff
@@ -0,0 +1,429 @@
+		if err != nil {
+			return err
+		}
+		return fmt.Errorf("unable to read GitLab approvel rule %s", d.Id())
+	}
+
+	d.Set("name", rule.Name)
+	d.SetPartial("name")
+	d.Set("approvals_required", rule.ApprovalsRequired)
+	d.SetPartial("approvals_required")
+
+	groupIDs := flattenApprovalRuleGroupIDs(rule.Groups)
+	if err := d.Set("group_ids", groupIDs); err != nil {
+		return err
+	}
+	d.SetPartial("group_ids")
+
+	eligibleApprovers := flattenApprovalRuleUserIDs(rule.EligibleApprovers)
+	if err := d.Set("eligible_approvers", eligibleApprovers); err != nil {
+		return err
+	}
+	d.SetPartial("eligible_approvers")
+
+	userIDs, err := getProjectApprovalRuleUserIDs(meta.(*gitlab.Client), d.Id())
+	if err != nil {
+		return err
+	}
+
+	if err := d.Set("user_ids", userIDs); err != nil {
+		return err
+	}
+
+	d.Partial(false)
+
+	return nil
+}
+
+func resourceGitlabProjectApprovalRuleUpdate(d *schema.ResourceData, meta interface{}) error {
+	var err error
+
+	if project, ruleID, err := splitApprovalRuleID(d.Id()); err == nil {
+		options := gitlab.UpdateProjectLevelRuleOptions{
+			Name:              gitlab.String(d.Get("name").(string)),
+			ApprovalsRequired: gitlab.Int(d.Get("approvals_required").(int)),
+			UserIDs:           expandApproverIds(d.GetOk("user_ids")),
+			GroupIDs:          expandApproverIds(d.GetOk("group_ids")),
+		}
+
+		ruleIDInt, err := strconv.Atoi(ruleID)
+		if err != nil {
+			return err
+		}
+
+		log.Printf("[DEBUG] Project %s update gitlab project-level approval rule %s", project, *options.Name)
+
+		client := meta.(*gitlab.Client)
+		if _, _, err = client.Projects.UpdateProjectApprovalRule(project, ruleIDInt, &options); err == nil {
+			return resourceGitlabProjectVariableRead(d, meta)
+		}
+	}
+
+	return err
+}
+
+func resourceGitlabProjectApprovalRuleDelete(d *schema.ResourceData, meta interface{}) error {
+	var err error
+
+	if project, ruleID, err := splitApprovalRuleID(d.Id()); err == nil {
+		ruleIDInt, err := strconv.Atoi(ruleID)
+		if err != nil {
+			return err
+		}
+
+		log.Printf("[DEBUG] Project %s delete gitlab project-level approval rule %d", project, ruleIDInt)
+
+		client := meta.(*gitlab.Client)
+		_, err = client.Projects.DeleteProjectApprovalRule(project, ruleIDInt)
+	}
+
+	return err
+}
+
+// getApprovalRuleByID checks the list of rules and finds the one that matches our rule ID.
+func getApprovalRuleByID(meta interface{}, id string) (*gitlab.ProjectApprovalRule, error) {
+	var rules []*gitlab.ProjectApprovalRule
+	var err error
+
+	projectID, ruleIDStr, err := splitApprovalRuleID(id)
+	if err != nil {
+		return nil, err
+	}
+
+	client := meta.(*gitlab.Client)
+	if rules, _, err = client.Projects.GetProjectApprovalRules(projectID); err == nil {
+		var ruleIDInt int
+
+		if ruleIDInt, err = strconv.Atoi(ruleIDStr); err != nil {
+			return nil, err
+		}
+
+		for _, r := range rules {
+			if r.ID == ruleIDInt {
+				log.Printf("[DEBUG] found project-level rule %+v", r)
+				return r, nil
+			}
+		}
+	}
```

Comment:
Uh oh. If the rule ID does not exist, then the for loop is going to finish without returning, which means you will return a `nil` rule and a `nil` error. Looking at the places where you call this function, I don't see you handling the `nil` rule case, so this is a bug.